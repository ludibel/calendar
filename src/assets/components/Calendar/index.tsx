import { FC, useEffect, useState, MouseEvent } from 'react';
import { Typography, IconButton, Divider, Card } from '@mui/material';
import Grid from '@mui/material/Unstable_Grid2'; // Grid version 2
import ArrowBackIosIcon from '@mui/icons-material/ArrowBackIos';
import ArrowForwardIosIcon from '@mui/icons-material/ArrowForwardIos';
// import interne
import { dateType } from '../../types';
import { dispos } from '../../datas/dispos';
import { convertDate, convertTime } from '../../services/convertionDate';
// import des composants mui stylés
import {
  StyledGridDate,
  StyledGridAvailability,
  StyledDivAvailabilityEmpty,
  StyledGridDatas,
  StyledDivAvailabilityDash,
  StyledGridContainer,
  StyledGridIcon,
  StyledButton,
  StyledGridMoreAvailability,
  StyledButtonMoreAvailability,
} from './styledCalendar';
const TITLEBUTTON = {
  MORE: "VOIR PLUS D'HORAIRE",
  LEST: "VOIR MOINS D'HORAIRE",
};

const Calendar: FC = () => {
  const [arraydatas, setArraydatas] = useState(dispos);
  // state pour la longueur du tableau de disponibilité
  const [arraylength, setArraylength] = useState(4);
  // state pour connaitre la valeur de la plus grande longueur de tableau slots
  const [greatValueArraySlots, setGreatValueArraySlots] = useState(4);
  // state pour voir le button voir plus d'horaire
  const [showButton, setShowButton] = useState(false);
  // nom pour le bouton plus ou moins d'horaire
  const [nameButton, setNameButton] = useState(TITLEBUTTON.MORE);

  // fonction permettant de récupérer la longueur du tableau des slots
  // et si cette longueur > 3 on affiche le button 'voir plus d'horaire"

  const getLengthArray = (array: Array<dateType>) => {
    const greatLength: number[] = [];
    {
      array.map((data) => {
        greatLength.push(data.slots.length);
        data.slots.length = Math.max(...greatLength);
        {
          data.slots.length <= 3 ? setShowButton(false) : setShowButton(true);
        }
        setGreatValueArraySlots(data.slots.length);
      });
    }
  };
  // fonction vérifiant la résolution de l'écran et selon la résolution modifie le tableau des dates
  const getResolution = () => {
    const resolution = window.screen.width;
    setArraylength(4);
    if (resolution < 768) {
      // En resolution <768 reduit le tableau à une longueur de 3 afin d'afficher que les 3 premiers jours
      // puis on mappe le tableau pour trouver la longueur du tableau de slots la plus grande entre les 3 jours
      setArraydatas(dispos.slice(0, 3));
      const newArray = dispos.slice(0, 3);
      getLengthArray(newArray);
    } else if (resolution < 1024) {
      // En resolution < 1024 reduit le tableau à une longueur de 4 afin d'afficher que les 4 premiers jours
      // puis on mappe le tableau pour trouver la longueur du tableau de slots la plus grande entre les 4 jour
      setArraydatas(dispos.slice(0, 4));
      const newArray = dispos.slice(0, 4);
      getLengthArray(newArray);
    } else {
      // on récupére 6 jours
      setArraydatas(dispos.slice(0, 6));
      const newArray = dispos.slice(0, 6);
      getLengthArray(newArray);
    }
  };

  useEffect(() => {
    // A chaque reinitilisation de la taille en effectuer le fonction getResolution()
    window.addEventListener('resize', getResolution);
    getResolution();
  }, [dispos]);

  const handleClickAvailabilityBefore = (
    event: MouseEvent<HTMLButtonElement>,
  ) => {
    console.log('jours précedents');
  };
  const handleClickAvailabilityAfter = (
    event: MouseEvent<HTMLButtonElement>,
  ) => {
    console.log('dispos suivantes');
  };

  const HandleMoreAvailability = () => {
    switch (nameButton) {
      case TITLEBUTTON.MORE:
        setArraylength(greatValueArraySlots);
        setNameButton(TITLEBUTTON.LEST);
        break;
      case TITLEBUTTON.LEST:
        setArraylength(4);
        setNameButton(TITLEBUTTON.MORE);
        break;
      default:
        setArraylength(greatValueArraySlots);
    }
  };

  return (
    <Card>
      <Grid
        container
        direction='row'
        justifyContent='center'
        alignContent='center'
        spacing={1}
      >
        <Grid xs={12} sm={4} md={3}>
          Profil
          <p>
            {' '}
            Lorem ipsum dolor sit amet consectetur, adipisicing elit. Sint eum a
            consequatur doloribus nemo incidunt exercitationem repellendus
            quisquam, adipisci quia!
          </p>
        </Grid>
        <Grid xs={12} sm={8} md={9}>
          <StyledGridContainer
            container
            direction='row'
            justifyContent='center'
            alignContent='center'
            columns={{ xs: 12, md: 14 }}
          >
            <StyledGridIcon xs={1}>
              <IconButton
                aria-label=' voie lesdates précedentes'
                onClick={handleClickAvailabilityBefore}
              >
                <ArrowBackIosIcon />
              </IconButton>
            </StyledGridIcon>
            {arraydatas.map((data, index) => (
              <StyledGridDatas xs={3} sm={2} key={`avaib-${data.day}`}>
                <StyledGridDate>
                  <Typography>
                    {convertDate(data.day, {
                      weekday: 'long',
                    })}
                  </Typography>
                  <Typography>
                    {convertDate(data.day, {
                      day: 'numeric',
                    })}{' '}
                    {convertDate(data.day, {
                      month: 'long',
                    }).substring(0, 3)}
                    .
                  </Typography>
                </StyledGridDate>
                <StyledGridAvailability>
                  {/* permet de créer un tableau de la longueur arraylength (4 en initial ou de la valeur calculée lors l'évenement onClick de "voir plus d'horaire")
                    et de mapper data.slots avec des index supérieurs à data.slots.length */}
                  {Array.from(new Array(arraylength)).map((_, i) => (
                    <Grid>
                      {data.slots[i] ? (
                        <StyledButton key={`slot-${index}`}>
                          {convertTime(data.slots[i], {
                            hour: '2-digit',
                            minute: '2-digit',
                          })}
                        </StyledButton>
                      ) : (
                        <StyledDivAvailabilityEmpty key={`slot-empty-${index}`}>
                          <StyledDivAvailabilityDash />
                        </StyledDivAvailabilityEmpty>
                      )}
                    </Grid>
                  ))}
                </StyledGridAvailability>
              </StyledGridDatas>
            ))}
            <StyledGridIcon xs={1}>
              <IconButton
                aria-label='voir les dates suivantes'
                onClick={handleClickAvailabilityAfter}
              >
                <ArrowForwardIosIcon />
              </IconButton>
            </StyledGridIcon>
          </StyledGridContainer>
          <Divider />
          {showButton ? (
            <StyledGridMoreAvailability>
              <StyledButtonMoreAvailability
                variant='text'
                onClick={HandleMoreAvailability}
              >
                {nameButton}
              </StyledButtonMoreAvailability>
            </StyledGridMoreAvailability>
          ) : (
            ''
          )}
        </Grid>
      </Grid>
    </Card>
  );
};

export default Calendar;
