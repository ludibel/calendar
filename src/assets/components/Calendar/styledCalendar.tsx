import styled from '@emotion/styled';

import { Box, Button } from '@mui/material';
import Grid from '@mui/material/Unstable_Grid2'; // Grid version 2

export const StyledBox = styled(Box)({
  backgroundColor: '#e5e5e5',
});
export const StyledGridDate = styled('div')({
  textAlign: 'center',
  marginBottom: '1em',
});
export const StyledGridAvailability = styled(Grid)({
  textAlign: 'center',
});
export const StyledDivAvailability = styled('div')({
  borderRadius: 4,
  fontSize: 14,
  fontWeight: 700,
  position: 'relative',
  backgroundColor: '#d4f4fe',
  marginTop: 10,
  marginBottom: 10,
  height: 34,
  lineHeight: '34px',
});
export const StyledDivAvailabilityEmpty = styled('div')({
  borderRadius: 1,
  width: '100%',
  height: 44,
  lineHeight: '34px',
  display: 'flex',
  alignItems: 'center',
  marginTop: 10,
  marginBottom: 10,
  marginRight: 10,
});
export const StyledDivAvailabilityDash = styled('div')({
  borderRadius: 1,
  backgroundColor: '#435f7133',
  width: 12,
  height: 2,
  marginRight: 'auto',
  marginLeft: 'auto',
});
export const StyledGridDatas = styled(Grid)({
  textAlign: 'center',
});
export const StyledBoxBis = styled(Box)(({ theme }) => ({
  backgroundColor: '#f6f6f6',
  marginRight: '3em',
  marginLeft: '3em',
}));
export const StyledGridContainer = styled(Grid)(({ theme }) => ({
  // display: 'flex',
}));
export const StyledGridIcon = styled(Grid)(({ theme }) => ({
  // flex: 0,
}));
export const StyledButton = styled(Button)(({ theme }) => ({
  borderRadius: 4,
  fontSize: 14,
  fontWeight: 700,
  position: 'relative',
  backgroundColor: '#d4f4fe',
  marginTop: 10,
  marginBottom: 10,
  height: 34,
  lineHeight: '34px',
  width: '100%',
  color: '#2b4660',
}));
export const StyledGridMoreAvailability = styled(Grid)(({ theme }) => ({
  textAlign: 'center',
}));
export const StyledButtonMoreAvailability = styled(Button)(({ theme }) => ({
  fontSize: '1em',
  fontWeight: 700,
  color: '#107aca',
}));
