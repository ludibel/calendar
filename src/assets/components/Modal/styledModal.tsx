import styled from '@emotion/styled';

import { Dialog, IconButton } from '@mui/material';

export const StyledDialog = styled(Dialog)({});

export const StyledIconButton = styled(IconButton)({
  position: 'absolute',
  right: 8,
  top: 8,
}) as typeof IconButton;
