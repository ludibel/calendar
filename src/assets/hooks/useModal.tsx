import { useState } from 'react';

// creation d'un hook useModal pour l'ouverture et la fermeture d'un composant modal
const useModal = () => {
  const [openModal, setOpenModal] = useState(false);

  const toggleOpenings = () => {
    setOpenModal(!openModal);
  };
  return {
    openModal,
    toggleOpenings,
  };
};
export default useModal;
