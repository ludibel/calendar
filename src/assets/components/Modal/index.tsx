import { FC } from 'react';
import { createPortal } from 'react-dom';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Typography,
} from '@mui/material';
import CloseIcon from '@mui/icons-material/Close';
import { StyledIconButton } from './styledModal';

export interface DialogTitleProps {
  id: string;
  children?: React.ReactNode;
  onClose: () => void;
}

export interface ModalProps {
  openModal: boolean;
  hide: () => void;
  modalContent: JSX.Element;
  headerTitle: string;
}

const DialogTitleWithIconClose = (props: DialogTitleProps) => {
  const { children, onClose } = props;
  return (
    <DialogTitle>
      {children}
      {onClose ? (
        <StyledIconButton
          aria-label='Fermer la fenÃªtre de dialogue'
          onClick={onClose}
        >
          <CloseIcon />
        </StyledIconButton>
      ) : null}
    </DialogTitle>
  );
};
// createPortal permet de rendre le composant en dehors du composant parent
const Modal: FC<ModalProps> = ({
  openModal,
  hide,
  modalContent,
  headerTitle,
}) =>
  openModal
    ? createPortal(
        <>
          <Dialog
            onClose={hide}
            aria-labelledby='customized-dialog-title'
            open={openModal}
            data-testid='modal'
          >
            <DialogTitleWithIconClose
              id='customized-dialog-title'
              onClose={hide}
            >
              {headerTitle}
            </DialogTitleWithIconClose>
            <DialogContent dividers>
              <Typography gutterBottom>{modalContent}</Typography>
            </DialogContent>
            <DialogActions>
              <Button autoFocus onClick={hide}>
                Fermer
              </Button>
            </DialogActions>
          </Dialog>
        </>,
        document.body,
      )
    : null;

export default Modal;
