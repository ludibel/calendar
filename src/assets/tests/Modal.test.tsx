import ReactDOM from 'react-dom';
import { Fragment } from 'react';
import {
  render,
  renderHook,
  act,
  screen,
  fireEvent,
} from '@testing-library/react';
import useModal from '../hooks/useModal';

import { vi } from 'vitest';

import Modal from '../components/Modal';
import Calendar from '../components/Calendar';

describe('Show modal', () => {
  //test de l'ouverture de la modal au clic sur le bouton
  test('open modalpenModal', () => {
    render(<Calendar />);
    fireEvent.click(screen.getByTestId('open-modal'));
    // TO DO expect
  });
});

describe('useModal', () => {
  // test que que toggleOpenings modifie bien le state openModal à true (car il est à false en initial)
  test('setopenModal change openModal', () => {
    const { result } = renderHook(() => useModal());

    act(() => {
      result.current.toggleOpenings();
    });

    expect(result.current.openModal).toBe(true);
  });
});

describe('Modal Component test with Portal', () => {
  // test que la Modal est bien rendu avec reactDOM portal on monte alors le portal avant de faire le test
  let mockHideModal: () => void;
  // avant on definit hide vi.fn va créer un espion sur une fonction
  beforeEach(() => {
    mockHideModal = vi.fn();
  });
  // Avant on crée le portal react
  beforeAll(() => {
    ReactDOM.createPortal = vi.fn((element) => {
      return element;
    });
  });
  // Après  on supprime le protal
  afterEach(() => {
    ReactDOM.createPortal.mockClear();
  });
  test('Show Modal correctly', () => {
    const { container } = render(
      <Modal
        openModal={true}
        hide={mockHideModal}
        headerTitle='Modal test'
        modalContent={
          <Fragment>
            Ton rendez-vous du XXX à HH:mm à bien éte enregistré
          </Fragment>
        }
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
