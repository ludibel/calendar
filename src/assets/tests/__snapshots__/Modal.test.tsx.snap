// Vitest Snapshot v1

exports[`Modal Component test with Portal > Show Modal correctly 1`] = `null`;
